In [1]:
!pip install -q pyomo

In [2]:
from pyomo.environ import * 

In [3]:
import numpy as np

**Ex 1 part 1**

Let $x_{ij}$ be the decision variable denoting the number of WMUs to be transported from warehouse $i$ to market $j$. Where $i$ takes values from set $\{1,2,3,4,5,6,7,8\}$ corresponding to warehouses at Ahmedabad, Bengaluru, Chennai, Delhi, Kolkata, Lucknow, Mumbai, Hyderabad respectively and $j$ takes values from set $\{1,2,3,4,5,6,7,8\}$ corresponding to markets at Hassan, Jodhpur, Trichy, Simla, vijayawada, Dibrugarh, Raipur, Agartala respectively.

It is clear that the the sum $\sum_{j=1}^{8} x_{ij}$ for all $i's$ denotes the number of WMUs transported from warehouse $i$. Similarly, $\sum_{i=1}^{8} x_{ij}$ for all $j's$ denotes no. of WMUs received at market $j$.

Hence, the corresponding linear program is,

Minimize $\sum_{i,j=1}^{8} C[i,j] x_{ij}$ \\
Sub. to. \\
$\sum_{j=1}^{8} x_{ij} \leq S_i  (\text{ available stock at warehouse i}) \  \forall i \\
\sum_{i=1}^{8} x_{ij} \leq D_i  (\text{ Demand at market j}) \  \forall j \\
And \ x_{ij} \ are \ nonnegative \  integers $

In [4]:
import pandas as pd

In [5]:
data_csvfile = pd.read_csv('lab5_ex1.csv')

In [6]:
data_csvfile.index

RangeIndex(start=0, stop=9, step=1)

In [7]:
print('num rows:', len(data_csvfile.index))

num rows: 9


In [8]:
print(data_csvfile.iloc[0,:]) # it displays cost of transportation from various warehouse to market at Ahmedabad

Warehouse-Market    Ahmedabad
Hassan                    627
Jodhpur                   617
Trichy                   1270
Simla                    2072
Vijaywada                1415
Dibrugarh                2943
Raipur                   1064
Agartala                 4265
Stocks                    185
Name: 0, dtype: object


In [9]:
data_csvfile.columns

Index(['Warehouse-Market', 'Hassan', 'Jodhpur', 'Trichy', 'Simla', 'Vijaywada',
       'Dibrugarh', 'Raipur', 'Agartala', 'Stocks'],
      dtype='object')

In [10]:
print('num  columns: ',len(data_csvfile.columns))

num  columns:  10


In [11]:
print(data_csvfile.iloc[:,1]) #it displays column content corresponding to Ahmedabad

0     627
1     159
2     609
3    2143
4    2742
5    2053
6     656
7     195
8     175
Name: Hassan, dtype: int64


In [12]:
# create a model
model_lab5_ex1 = ConcreteModel()

In [13]:
# we can set certain values using python variables
M = len(data_csvfile.index)-1 #we declare a Python variable M denoting the number of warehouses 
print('num of warehouses: ',M)

num of warehouses:  8


In [14]:
N = len(data_csvfile.columns)-2 #we declare a Python variable N denoting the number of markets
print('num of markets: ',N)

num of markets:  8


In [15]:
row_indices = range(M)
col_indices = range(N)

print(row_indices)
print(col_indices)

range(0, 8)
range(0, 8)


In [16]:
#to get the contents of costs alone from the file we use the following
for row_idx in row_indices:
  for col_idx in col_indices:
    print('cost[',row_idx,'][',col_idx+1,']:',data_csvfile.iloc[row_idx, col_idx+1])

cost[ 0 ][ 1 ]: 627
cost[ 0 ][ 2 ]: 617
cost[ 0 ][ 3 ]: 1270
cost[ 0 ][ 4 ]: 2072
cost[ 0 ][ 5 ]: 1415
cost[ 0 ][ 6 ]: 2943
cost[ 0 ][ 7 ]: 1064
cost[ 0 ][ 8 ]: 4265
cost[ 1 ][ 1 ]: 159
cost[ 1 ][ 2 ]: 1432
cost[ 1 ][ 3 ]: 351
cost[ 1 ][ 4 ]: 3372
cost[ 1 ][ 5 ]: 237
cost[ 1 ][ 6 ]: 3373
cost[ 1 ][ 7 ]: 1456
cost[ 1 ][ 8 ]: 5059
cost[ 2 ][ 1 ]: 609
cost[ 2 ][ 2 ]: 1943
cost[ 2 ][ 3 ]: 77
cost[ 2 ][ 4 ]: 3365
cost[ 2 ][ 5 ]: 298
cost[ 2 ][ 6 ]: 3021
cost[ 2 ][ 7 ]: 1484
cost[ 2 ][ 8 ]: 4803
cost[ 3 ][ 1 ]: 2143
cost[ 3 ][ 2 ]: 671
cost[ 3 ][ 3 ]: 3876
cost[ 3 ][ 4 ]: 537
cost[ 3 ][ 5 ]: 1902
cost[ 3 ][ 6 ]: 1941
cost[ 3 ][ 7 ]: 1682
cost[ 3 ][ 8 ]: 3348
cost[ 4 ][ 1 ]: 2742
cost[ 4 ][ 2 ]: 2089
cost[ 4 ][ 3 ]: 1931
cost[ 4 ][ 4 ]: 2199
cost[ 4 ][ 5 ]: 1226
cost[ 4 ][ 6 ]: 739
cost[ 4 ][ 7 ]: 691
cost[ 4 ][ 8 ]: 1128
cost[ 5 ][ 1 ]: 2053
cost[ 5 ][ 2 ]: 860
cost[ 5 ][ 3 ]: 2995
cost[ 5 ][ 4 ]: 68
cost[ 5 ][ 5 ]: 2069
cost[ 5 ][ 6 ]: 2065
cost[ 5 ][ 7 ]: 993
cost[ 5 ][ 8 ]: 3342
cost[ 6 ]

In [17]:
#declare the decision variables in the model
model_lab5_ex1.x = Var(row_indices, col_indices, domain= NonNegativeIntegers)

In [18]:
model_lab5_ex1.objective = Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab5_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

In [19]:
model_lab5_ex1.constraints = ConstraintList()

In [20]:
print(data_csvfile.iloc[:8,N+1])

0    185
1    350
2    405
3    280
4    105
5     75
6    425
7    270
Name: Stocks, dtype: int64


In [21]:
print(data_csvfile.iloc[M,1:N+1])

Hassan       175
Jodhpur      400
Trichy       550
Simla         70
Vijaywada    650
Dibrugarh    100
Raipur        65
Agartala      85
Name: 8, dtype: object


In [22]:
#stock constraints for warehouses
for i in row_indices:
  model_lab5_ex1.constraints.add(expr=sum(model_lab5_ex1.x[i,j] for j in col_indices) <= data_csvfile.iloc[i,9]) 


In [23]:
#demand constraint for markets
for j in col_indices:
  model_lab5_ex1.constraints.add(expr=sum(model_lab5_ex1.x[i,j] for i in row_indices) == data_csvfile.iloc[M,j+1])


In [24]:
model_lab5_ex1.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :   64 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7)}
    x_index_0 : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
       

In [25]:
!apt-get install -y -qq coinor-cbc

In [26]:
opt_cbc = SolverFactory('cbc')

In [27]:
result = opt_cbc.solve(model_lab5_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [28]:
# display solution
print('\nObjective = ', model_lab5_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Num of WMUs between ',data_csvfile['Warehouse-Market'][i],' and ', data_csvfile.columns[j+1],':', model_lab5_ex1.x[i,j].value)

print('\nConstraints')
model_lab5_ex1.constraints.display()


Objective =  1166465.0

Decision Variables
Num of WMUs between  Ahmedabad  and  Hassan : 0.0
Num of WMUs between  Ahmedabad  and  Jodhpur : 185.0
Num of WMUs between  Ahmedabad  and  Trichy : 0.0
Num of WMUs between  Ahmedabad  and  Simla : 0.0
Num of WMUs between  Ahmedabad  and  Vijaywada : 0.0
Num of WMUs between  Ahmedabad  and  Dibrugarh : 0.0
Num of WMUs between  Ahmedabad  and  Raipur : 0.0
Num of WMUs between  Ahmedabad  and  Agartala : 0.0
Num of WMUs between  Bengaluru  and  Hassan : 0.0
Num of WMUs between  Bengaluru  and  Jodhpur : 0.0
Num of WMUs between  Bengaluru  and  Trichy : 0.0
Num of WMUs between  Bengaluru  and  Simla : 0.0
Num of WMUs between  Bengaluru  and  Vijaywada : 350.0
Num of WMUs between  Bengaluru  and  Dibrugarh : 0.0
Num of WMUs between  Bengaluru  and  Raipur : 0.0
Num of WMUs between  Bengaluru  and  Agartala : 0.0
Num of WMUs between  Chennai  and  Hassan : 0.0
Num of WMUs between  Chennai  and  Jodhpur : 0.0
Num of WMUs between  Chennai  and  Tric

**Ex 1 part 7**

185 WMUs are to be transported from Ahmedabad to Jodhpur \\
350 WMUs are to be transported from Benguluru to Vijaywada \\
375 WMUs are to be transported from Chennai to Vijaywada \\
30 WMUs are to be transported from Chennai to Vijaywada \\
200 WMUs are to be transported from Delhi to Jodhpur \\
80 WMUs are to be transported from Delhi to Dibrugarh \\
20 WMUs are to be transported from kolkata to Dibrugarh \\
85 WMUs are to be transported from Kolkata to Agartala \\
70 WMUs are to be transported from Lucknow to Simla \\
5 WMUs are to be transported from Lucknow to Raipur \\
175 WMUs are to be transported from Mumbai to Hassan \\
15 WMUs are to be transported from Mumbai to Jodhpur \\
175 WMUs are to be transported from Mumbai to Trichy \\
60 WMUs are to be transported from Mumbai to Raipur \\
270 WMUs are to be transported from Hyderabad to Vijaywada \\
 \\
And the the cost will be 1166465

**Ex 1 part 8**

We can assign a very very high transportation cost for movement from benguluru to Hassan in the csv file.
Hence the LPP algorith will automaticaly assign 0 for that route.

In [29]:
data_csvfile.loc[1,'Hassan']=150000

In [30]:
print(data_csvfile)

  Warehouse-Market  Hassan  Jodhpur  ...  Raipur  Agartala  Stocks
0        Ahmedabad     627      617  ...    1064      4265     185
1        Bengaluru  150000     1432  ...    1456      5059     350
2          Chennai     609     1943  ...    1484      4803     405
3            Delhi    2143      671  ...    1682      3348     280
4          Kolkata    2742     2089  ...     691      1128     105
5          Lucknow    2053      860  ...     993      3342      75
6           Mumbai     656      817  ...    1137      3887     425
7        Hyderabad     195     1862  ...     563      4441     270
8           Demand     175      400  ...      65        85    2095

[9 rows x 10 columns]


In [31]:
model_lab5_ex1.new_objective = Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab5_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

In [32]:
model_lab5_ex1.objective.deactivate()

In [33]:
model_lab5_ex1.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :   64 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7)}
    x_index_0 : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
       

In [34]:
result = opt_cbc.solve(model_lab5_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [35]:
# display solution
print('\nObjective = ', model_lab5_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Num of WMUs between ',data_csvfile['Warehouse-Market'][i],' and ', data_csvfile.columns[j+1],':', model_lab5_ex1.x[i,j].value)

print('\nConstraints')
model_lab5_ex1.constraints.display()


Objective =  1166465.0

Decision Variables
Num of WMUs between  Ahmedabad  and  Hassan : 0.0
Num of WMUs between  Ahmedabad  and  Jodhpur : 185.0
Num of WMUs between  Ahmedabad  and  Trichy : 0.0
Num of WMUs between  Ahmedabad  and  Simla : 0.0
Num of WMUs between  Ahmedabad  and  Vijaywada : 0.0
Num of WMUs between  Ahmedabad  and  Dibrugarh : 0.0
Num of WMUs between  Ahmedabad  and  Raipur : 0.0
Num of WMUs between  Ahmedabad  and  Agartala : 0.0
Num of WMUs between  Bengaluru  and  Hassan : 0.0
Num of WMUs between  Bengaluru  and  Jodhpur : 0.0
Num of WMUs between  Bengaluru  and  Trichy : 0.0
Num of WMUs between  Bengaluru  and  Simla : 0.0
Num of WMUs between  Bengaluru  and  Vijaywada : 350.0
Num of WMUs between  Bengaluru  and  Dibrugarh : 0.0
Num of WMUs between  Bengaluru  and  Raipur : 0.0
Num of WMUs between  Bengaluru  and  Agartala : 0.0
Num of WMUs between  Chennai  and  Hassan : 0.0
Num of WMUs between  Chennai  and  Jodhpur : 0.0
Num of WMUs between  Chennai  and  Tric

**Ex 1 part 9**

185 WMUs are to be transported from Ahmedabad to Jodhpur \\
350 WMUs are to be transported from Benguluru to Vijaywada \\
375 WMUs are to be transported from Chennai to Vijaywada \\
30 WMUs are to be transported from Chennai to Vijaywada \\
200 WMUs are to be transported from Delhi to Jodhpur \\
80 WMUs are to be transported from Delhi to Dibrugarh \\
20 WMUs are to be transported from kolkata to Dibrugarh \\
85 WMUs are to be transported from Kolkata to Agartala \\
70 WMUs are to be transported from Lucknow to Simla \\
5 WMUs are to be transported from Lucknow to Raipur \\
175 WMUs are to be transported from Mumbai to Hassan \\
15 WMUs are to be transported from Mumbai to Jodhpur \\
175 WMUs are to be transported from Mumbai to Trichy \\
60 WMUs are to be transported from Mumbai to Raipur \\
270 WMUs are to be transported from Hyderabad to Vijaywada \\
 \\
And the the cost will be 1166465

There is no change in the solution as the the no. of WMUs to be transported from Benguluru to Hassan was already zero.

**Ex 1 part 10**

In [36]:
data_csvfile.loc[5,'Simla']=150000
data_csvfile.loc[4,'Agartala']=150000

In [37]:
print(data_csvfile)

  Warehouse-Market  Hassan  Jodhpur  ...  Raipur  Agartala  Stocks
0        Ahmedabad     627      617  ...    1064      4265     185
1        Bengaluru  150000     1432  ...    1456      5059     350
2          Chennai     609     1943  ...    1484      4803     405
3            Delhi    2143      671  ...    1682      3348     280
4          Kolkata    2742     2089  ...     691    150000     105
5          Lucknow    2053      860  ...     993      3342      75
6           Mumbai     656      817  ...    1137      3887     425
7        Hyderabad     195     1862  ...     563      4441     270
8           Demand     175      400  ...      65        85    2095

[9 rows x 10 columns]


In [38]:
model_lab5_ex1.objective3 = Objective(expr=sum(data_csvfile.iloc[i,j+1]*model_lab5_ex1.x[i,j] for i in row_indices for j in col_indices),sense=minimize)

In [39]:
model_lab5_ex1.new_objective.deactivate()

In [40]:
model_lab5_ex1.pprint()

4 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   16 : {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain              : Size : Members
        None :     2 : x_index_0*x_index_1 :   64 : {(0, 0), (0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (0, 6), (0, 7), (1, 0), (1, 1), (1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (2, 0), (2, 1), (2, 2), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (3, 0), (3, 1), (3, 2), (3, 3), (3, 4), (3, 5), (3, 6), (3, 7), (4, 0), (4, 1), (4, 2), (4, 3), (4, 4), (4, 5), (4, 6), (4, 7), (5, 0), (5, 1), (5, 2), (5, 3), (5, 4), (5, 5), (5, 6), (5, 7), (6, 0), (6, 1), (6, 2), (6, 3), (6, 4), (6, 5), (6, 6), (6, 7), (7, 0), (7, 1), (7, 2), (7, 3), (7, 4), (7, 5), (7, 6), (7, 7)}
    x_index_0 : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
       

In [41]:
result = opt_cbc.solve(model_lab5_ex1)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [42]:
# display solution
print('\nObjective = ', model_lab5_ex1.objective())

print('\nDecision Variables')
for i in row_indices:
  for j in col_indices:
    print('Num of WMUs between ',data_csvfile['Warehouse-Market'][i],' and ', data_csvfile.columns[j+1],':', model_lab5_ex1.x[i,j].value)

print('\nConstraints')
model_lab5_ex1.constraints.display()


Objective =  1289875.0

Decision Variables
Num of WMUs between  Ahmedabad  and  Hassan : 0.0
Num of WMUs between  Ahmedabad  and  Jodhpur : 185.0
Num of WMUs between  Ahmedabad  and  Trichy : 0.0
Num of WMUs between  Ahmedabad  and  Simla : 0.0
Num of WMUs between  Ahmedabad  and  Vijaywada : 0.0
Num of WMUs between  Ahmedabad  and  Dibrugarh : 0.0
Num of WMUs between  Ahmedabad  and  Raipur : 0.0
Num of WMUs between  Ahmedabad  and  Agartala : 0.0
Num of WMUs between  Bengaluru  and  Hassan : 0.0
Num of WMUs between  Bengaluru  and  Jodhpur : 0.0
Num of WMUs between  Bengaluru  and  Trichy : 0.0
Num of WMUs between  Bengaluru  and  Simla : 0.0
Num of WMUs between  Bengaluru  and  Vijaywada : 350.0
Num of WMUs between  Bengaluru  and  Dibrugarh : 0.0
Num of WMUs between  Bengaluru  and  Raipur : 0.0
Num of WMUs between  Bengaluru  and  Agartala : 0.0
Num of WMUs between  Chennai  and  Hassan : 0.0
Num of WMUs between  Chennai  and  Jodhpur : 0.0
Num of WMUs between  Chennai  and  Tric

**Ex 1 part 11**

185 WMUs are to be transported from Ahmedabad to Jodhpur \\
350 WMUs are to be transported from Benguluru to Vijaywada \\
375 WMUs are to be transported from Chennai to Vijaywada \\
30 WMUs are to be transported from Chennai to Vijaywada \\
200 WMUs are to be transported from Delhi to Jodhpur \\
70 WMUs are to be transported from Delhi to Simla \\
10 WMUs are to be transported from Delhi to Agartala \\

100 WMUs are to be transported from kolkata to Dibrugarh \\
5 WMUs are to be transported from Kolkata to Raipur \\
75 WMUs are to be transported from Lucknow to Agartala \\
175 WMUs are to be transported from Mumbai to Hassan \\
15 WMUs are to be transported from Mumbai to Jodhpur \\
175 WMUs are to be transported from Mumbai to Trichy \\
60 WMUs are to be transported from Mumbai to Raipur \\
270 WMUs are to be transported from Hyderabad to Vijaywada \\
 \\
And the the cost will be 1289875

Here we are getting more optimal value as we have ignored the given three routes(of which only last 2 are affecting the solution)